pkg/common/cns-lib/volume/manager.go

```diff
@@ -565,63 +576,184 @@ func (m *defaultManager) DeleteVolume(ctx context.Context, volumeID string, dele
+	if err != nil {
+		if cnsvsphere.IsNotFoundError(err) {
+			log.Infof("VolumeID: %q, not found, thus returning success", volumeID)
+			return nil
+		}
+		log.Errorf("CNS DeleteVolume failed from the  vCenter %q with err: %v", m.virtualCenter.Config.Host, err)
+		return err
+	}
+	// Get the taskInfo.
+	taskInfo, err := cns.GetTaskInfo(ctx, task)
+	if err != nil || taskInfo == nil {
+		log.Errorf("failed to get DeleteVolume taskInfo from vCenter %q with err: %v",
+			m.virtualCenter.Config.Host, err)
+		return err
+	}
+	log.Infof("DeleteVolume: volumeID: %q, opId: %q", volumeID, taskInfo.ActivationId)
+	// Get the task results for the given task.
+	taskResult, err := getTaskResultFromTaskInfo(ctx, taskInfo)
+	if err != nil {
+		log.Errorf("failed to get DeleteVolume task result with error: %v", err)
+		return err
+	}
+	volumeOperationRes := taskResult.GetCnsVolumeOperationResult()
+	if volumeOperationRes.Fault != nil {
+		return logger.LogNewErrorf(log, "failed to delete volume: %q, fault: %q, opID: %q",
+			volumeID, spew.Sdump(volumeOperationRes.Fault), taskInfo.ActivationId)
+	}
+	log.Infof("DeleteVolume: Volume deleted successfully. volumeID: %q, opId: %q",
+		volumeID, taskInfo.ActivationId)
+	return nil
+}
+
+// deleteVolumeWithIdempotency attempts to delete the volume on CNS. CNS task information is persisted by leveraging
+// the VolumeOperationRequest interface.
+func (m *defaultManager) deleteVolumeWithIdempotency(ctx context.Context, volumeID string, deleteDisk bool) error {
+	log := logger.GetLogger(ctx)
+	var (
+		// Reference to the DeleteVolume task on CNS.
+		task *object.Task
+		// Name of the CnsVolumeOperationRequest instance.
+		instanceName = "delete-" + volumeID
+		// Local instance of DeleteVolume details that needs to
+		// be persisted.
+		volumeOperationDetails *cnsvolumeoperationrequest.VolumeOperationRequestDetails
+	)
+	if m.operationStore == nil {
+		return logger.LogNewError(log, "operation store cannot be nil")
+	}
+	// Determine if CNS needs to be invoked.
+	volumeOperationDetails, err := m.operationStore.GetRequestDetails(ctx, instanceName)
+	switch {
+	case err == nil:
+		if volumeOperationDetails.OperationDetails != nil {
+			// Validate if previous attempt was
+			// successful.
+			if volumeOperationDetails.OperationDetails.TaskStatus == taskInvocationStatusSuccess {
+				return nil
+			}
+			// Validate if previous operation is
+			// pending.
+			if volumeOperationDetails.OperationDetails.TaskStatus == taskInvocationStatusInProgress &&
+				volumeOperationDetails.OperationDetails.TaskID != "" {
+				taskMoRef := vim25types.ManagedObjectReference{
+					Type:  "Task",
+					Value: volumeOperationDetails.OperationDetails.TaskID,
+				}
+				task = object.NewTask(m.virtualCenter.Client.Client, taskMoRef)
+			}
+		}
+	case apierrors.IsNotFound(err):
+		volumeOperationDetails = createRequestDetails(instanceName, "", "", 0, metav1.Now(),
```

Comment:
This operation is not thread-safe. Can we have a race between two delete requests for the same volume?

Reply:
This operation itself doesnt need to be thread-safe, it only creates an in-memory instance of the object. This object is modified and later stored on the API server. 
But in general, once we remove the gocsi code, it looks like we will need some serialization logic. Here's more info - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/898/files#r644224790

Since gocsi is still enabled, im taking this as a follow up PR as i need to think about where/how to handle serialization in these cases.

Reply:
Makes sense.